src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,934 @@
+            })
+            .unwrap()
+    }
+
+    fn update_statistics(&mut self) {
+        if self.last_target_file.is_some() {
+            SCHED_L0_TARGET_FLOW_GAUGE.set(self.l0_target_flow as i64);
+        } else {
+            SCHED_L0_TARGET_FLOW_GAUGE.set(0);
+        }
+
+        if let Some(throttle_cf) = self.throttle_cf.as_ref() {
+            SCHED_THROTTLE_CF_GAUGE
+                .with_label_values(&[throttle_cf])
+                .set(1);
+            for (cf, _) in &self.cf_checkers {
+                if cf != throttle_cf {
+                    SCHED_THROTTLE_CF_GAUGE.with_label_values(&[cf]).set(0);
+                }
+            }
+        } else {
+            for (cf, _) in &self.cf_checkers {
+                SCHED_THROTTLE_CF_GAUGE.with_label_values(&[cf]).set(0);
+            }
+        }
+
+        let rate =
+            self.limiter.total_bytes_consumed() as f64 / self.last_record_time.elapsed_secs();
+        if self.limiter.total_bytes_consumed() != 0 {
+            self.recorder.observe(rate as u64);
+        }
+        SCHED_WRITE_FLOW_GAUGE.set(rate as i64);
+        self.last_record_time = Instant::now_coarse();
+        self.limiter.reset_statistics();
+    }
+
+    fn adjust_pending_compaction_bytes(&mut self, cf: String) {
+        let hard = (self.pending_compaction_bytes_hard_limit as f64).log2();
+        let soft = (self.pending_compaction_bytes_soft_limit as f64).log2();
+
+        let num = (self
+            .engine
+            .kv_engine()
+            .get_cf_pending_compaction_bytes(&cf)
+            .unwrap_or(None)
+            .unwrap_or(0) as f64)
+            .log2();
+        let checker = self.cf_checkers.get_mut(&cf).unwrap();
+        checker
+            .long_term_pending_bytes
+            .observe((num * RATIO_PRECISION) as u64);
+        SCHED_PENDING_COMPACTION_BYTES_GAUGE
+            .with_label_values(&[&cf])
+            .set(checker.long_term_pending_bytes.get_avg() as i64);
+        if checker.on_start_pending_bytes {
+            if num < soft {
+                checker.on_start_pending_bytes = false;
+            } else if checker.long_term_pending_bytes.trend() == Trend::Increasing {
+                // the write is accumulating, still need to throttle
+                checker.on_start_pending_bytes = false;
+            } else {
+                // still on start, should not throttle now
+                return;
+            }
+        }
+
+        let pending_compaction_bytes = checker.long_term_pending_bytes.get_avg() / RATIO_PRECISION;
+        drop(checker);
+
+        for (_, checker) in &self.cf_checkers {
+            if num < (checker.long_term_pending_bytes.get_recent() as f64) / RATIO_PRECISION {
+                return;
+            }
+        }
+
+        let ratio = if pending_compaction_bytes < soft {
+            0
+        } else {
+            let new_ratio = (pending_compaction_bytes - soft) / (hard - soft);
```

Comment:
`Say if the pending bytes just spike for a short period of time, no matter how large the value gets, maybe we don't need to throttle a lot.`

That's why I use a large moving window to smooth pending bytes and logarithm the pending bytes. It would avoid the spike of it.